### PR TITLE
Fix shader compilation errors with Adreno GPUs, Implement N64 blending using framebuffer fetch extensions

### PR DESCRIPTION
**Pull request by fzurita**
Adreno GPUs don't like it when sampler2D is passed as a function argument. Also removed comments from what we pass to the shader compiler to reduce string size. Some GPU drivers have limits on shader size.

### CODE REVIEW

**Comment by fzurita** [comment]
Don't merge yet. There are still some issues with Adreno with "Hybrid filter" is disabled. It seems to work ok with it enabled for now.

**Comment by fzurita** [comment]
Ok, I think this is ok to be merged. It seems to be a separate issue unrelated to this that is causing Adreno GPUs to crash.

**Comment by fzurita** [comment]
It looks like c48841cfebff118af634135bc0f6a6a02bf4e7d4 broke Adreno devices.

**Comment by fzurita** [comment]
I went ahead and disabled dual source blending for Adreno devices under this same pull request.

**Comment by oddMLan** [comment]
I heard there is an alternative to Dual Source Blending for mobile devices that don't support it...

**Comment by fzurita** [comment]
Yeah, there is. Dolphin implemented it.

**Comment by fzurita** [comment]
https://github.com/dolphin-emu/dolphin/pull/6141

For reference, that is the dolphin implementation that doesn't require the extension.

**Comment by fzurita** [comment]
And it looks like @standard-two-simplex already provided a solution here: https://github.com/gonetz/GLideN64/issues/1647

It's probably worth implementing.

**Comment by fzurita** [comment]
Also, dolphin also had to workaround a Adreno driver bug, it's probably what we are seeing here.

**Comment by fzurita** [comment]
Dolphin workaround doesn't work for us for some reason. The GPU driver still crashes.

**Comment by gonetz** [comment]
@oddMLan could you please review these changes? You are the author of hybrid filter code.

**Comment by fzurita** [comment]
I added the framebuffer fetch implementation of dual source blending work the help of @standard-two-simplex. This works with Adreno and Mali GPUs:
gauntlet_legends-000
pokemon_stadium-000 (1)

**Comment by fzurita** [comment]
@standard-two-simplex Can you make sure that I didn't break dual source blending on PCs?

**Comment by standard-two-simplex** [comment]
Yes, I'll check everything later today.

**Comment by standard-two-simplex** [comment]
The math seems correct and I can run the dual_source_blending part in my computer without issues, so nothing seems broken.

I do not see the need to create special output `realFragColor` and then copy `fragColor` into it. Is it ok if I add some cleanup and abstract all extension specific code to the shader fragment header? This way the body would be the same for all cases.

**Comment by fzurita** [comment]
The reason I have ```realFragColor``` is because of this: https://github.com/dolphin-emu/dolphin/pull/6141/commits/8d68adcaf3442122fff5bb0988c2f06b0128f021

I guess I could try removing it and see if it works correctly without it since I was playing it safe from the start.

Feel free to clean it up and I can test your changes.

**Comment by standard-two-simplex** [comment]
@fzurita I created a PR to you branch.

I think you can remove `ShaderFragmentMainEndSpecial()` as well.

I would also recommend renaming f3ce000 to "Implement N64 blending using framebuffer fetch extensions" as it has nothing to do with dual source blending (which means outputing two colors/alphas from fragment shader and using them in OpenGLs blender).

**Comment by fzurita** [comment]
@standard-two-simplex Thanks. I left you a question in your pull request.

**Comment by oddMLan** [comment]
> @oddMLan could you please review these changes? You are the author of hybrid filter code.

Hybrid filter seems to work fine on my Win10 destkop with AMD gfx and so does dual source blending
GLideN64_POKEMON_STADIUM_000

**Comment by fzurita** [comment]
@standard-two-simplex I had to make a few minor corrections for shader compilation errors. Other than that, it works correctly with Adreno and Mali GPUs.

**Comment by fzurita** [comment]
I updated the pull request with your changes.

File: src/CMakeLists.txt [hunk 1]
```diff
@@ -439,6 +439,10 @@ if(GL_PROFILE)
     )
 endif(GL_PROFILE)
 
+add_definitions(
+        -DGL_DEBUG
+)
+
```
Comment (gonetz) [inline src/CMakeLists.txt:445]: Why?

Comment (fzurita) [reply src/CMakeLists.txt:445]: Whoops, I was using that while testing. I'll take it off.

File: src/GraphicsDrawer.cpp [hunk 2]
```diff
@@ -23,6 +23,7 @@
 #include "RSP.h"
 #include "RDP.h"
 #include "VI.h"
+#include "Log.h"
```
Comment (gonetz) [inline src/GraphicsDrawer.cpp:26]: Unnecessary change.

File: src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp [hunk 3]
```diff
@@ -1706,10 +1730,10 @@ class ShaderFragmentRenderTarget : public ShaderPart
 	}
 };
 
-class ShaderFragmentMainEnd : public ShaderPart
+class ShaderFragmentMainEndSpecial : public ShaderPart
```
Comment (fzurita) [inline src/Graphics/OpenGLContext/GLSL/glsl_CombinerProgramBuilder.cpp:1733]: @gonetz I forgot to remove this class. This is not needed.

**Comment by fzurita** [comment]
@standard-two-simplex unfortunately, we have some issues with this. Not sure how many games are affected bit at least turok and blast corps are.

Blast corps has flickering textures. Turok text is all messed up, see this:
turok_dinosaur_hunte-000

Any ideas?
